### PR TITLE
feat: associate templates with workspaces

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

- break up `det template set` into `det template create|set-value`
- maintain old set behavior
- add an option to declarative argparse to mark options and commands as deprecated.
  -  warn on usage and on --help
- deprecate old template put endpoint and `det template set`

DET-9452 DET-9335

## Test Plan

- added crud tests for template apis

manual testing:
- create, update,  list, describe, and delete templates
- try creating with a specifying a specific workspace name and check the workspace id is persisted correctly. (check the api response for now or use det tpl ls)

## Commentary (optional)

maybe this is a another way we can updated the cli options:
https://github.com/determined-ai/determined/blob/81c57d06b42820c60053d25ec254a28cff1a4e41/harness/determined/cli/experiment.py#L1250-L1254

TODO:
- [x] api codes
- [x] depends on #6572 

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: harness/determined/cli/template.py [hunk 1]
```diff
@@ -42,14 +43,47 @@ def describe_template(args: Namespace) -> None:
...
+        name=args.template_name, config=body, workspaceId=workspace_id
+    )
+    bindings.post_PostTemplate(
+        cli.setup_session(args), template_name=args.template_name, body=v1_template
+    )
+    print(colored("Created template {}".format(args.template_name), "green"))
+
+
+@authentication.required
+def patch_tempalte_config(args: Namespace) -> None:
```
Comment (erikwilson) [inline harness/determined/cli/template.py:77]:
```suggestion
def patch_template(args: Namespace) -> None:
```

File: harness/determined/cli/template.py [hunk 2]
```diff
@@ -64,14 +98,25 @@ def remove_templates(args: Namespace) -> None:
             Arg("-d", "--details", action="store_true",
                 help="show the configs of the templates"),
         ], is_default=True),
+        Cmd("set", set_template, "set config template", [
+            Arg("template_name", help="template name"),
+            Arg("template_file", type=FileType("r"),
+                help="config template file (.yaml)"),
+        ], deprecation_message="use other options offered by `det template` instead"),
```
Comment (erikwilson) [inline harness/determined/cli/template.py:105]: nit: should we explicitly call out create & update here?

**Comment by erikwilson** [comment]
overall looks good, appears that CI has some linting issues. might be good to add a manual test plan for release testing.

File: harness/determined/cli/template.py [hunk 3]
```diff
@@ -64,14 +98,25 @@ def remove_templates(args: Namespace) -> None:
...
                 Arg("template_name", type=str, help="template name"),
             ]),
-        Cmd("set", set_template, "set config template", [
+        Cmd("create", create_template, "create config template", [
+            Arg("template_name", help="template name"),
+            Arg("template_file", type=FileType("r"),
+                help="config template file (.yaml)"),
+            workspace_arg,
+        ]),
+        Cmd("update-config", patch_tempalte_config, "update config template", [
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply harness/determined/cli/template.py:116]: hmm yeah, that would work. I added that to keep it explicit considering we might have other attributes to update, maybe a workspace id later on.
trying to update multiple attributes with a single CLI call in a patch fashion gets weird. maybe there is a good pattern we should be using for that

File: harness/determined/cli/template.py (hunk 1 above)
Comment (hamidzr) [reply harness/determined/cli/template.py:77]: we should keep this in sync with whatever we do https://github.com/determined-ai/determined/pull/6605#discussion_r1182774437 I'll update if we decide to rename the cli cmd to just `template update`

File: harness/determined/common/declarative_argparse.py [hunk 4]
```diff
@@ -30,6 +33,32 @@ def generate_aliases(spec: str) -> Tuple[str, List[str]]:
...
+
+
+def deprecation_warning(message: str, color: bool = True) -> str:
+    msg = f"DEPRECATED: {message}"
+    return colored(msg, "yellow") if color else msg
+
+
+def warn_on_usage(message: str) -> Callable:
+    """
+    Decorator to prints a deprecation warning when the wrapped function is
```
Comment (hamidzr) [inline harness/determined/common/declarative_argparse.py:47]: typo

File: master/internal/api_templates.go [hunk 5]
```diff
@@ -42,16 +43,62 @@ func (a *apiServer) GetTemplate(
...
+	); err {
+	case nil:
+		return &res, nil
+	default:
+		return nil, status.Errorf(codes.Internal, "error posting template %s to db: %s",
+			req.Template.Name, err.Error())
+	}
+}
+
+func (a *apiServer) PatchTemplateConfig(
```
Comment (hamidzr) [inline master/internal/api_templates.go:83]: should this be designed more as a generic patch endpoint which for now just allows setting the config? communicating that through protobuf could be tricky

**Review by erikwilson — APPROVED** [review]

**Comment by hamidzr** [comment]
what do we think about `det template set config <file>` inline with det e and det command

**Comment by hamidzr** [comment]
thanks for the reviews @erikwilson

**Pull request merged**